### DISCUSSION TITLE
Visualizing/consuming code coverage data

### DISCUSSION THREAD

**Discussion by sberrevoets** [MobileNativeFoundation/discussions#180] [General]
Code coverage (i.e. how much code is covered by automated tests) is surprisingly complicated, with tools whose UX never really seem to hit the mark.

[`llvm-cov`](https://llvm.org/docs/CommandGuide/llvm-cov.html) and [`genhtml`](https://linux.die.net/man/1/genhtml) (part of [`lcov`](https://github.com/linux-test-project/lcov)) have the option to generate source overlays in HTML format, but that HTML looks like it's straight from the '90s. [JaCoCo](https://www.jacoco.org/jacoco/) is a Java-only library with basic/limited support for Kotlin and produces similarly outdated-looking HTML pages. None of these tools provide any kind of customization. There's also JSON output, but then it doesn't include source overlays to show which lines of code aren't covered, and you have to build some UI to navigate the vast amounts of data.

Xcode has an internal UI for visualizing code coverage, but if you're using Bazel and want hermetic builds this integration doesn't work.

We've tried codecov in the past but it wasn't very happy with the amount of code + number of modules we were dealing with.

It seems like, although controversial, code coverage is still the primary way of indicating how well tested a codebase is, yet none of the tools around it optimize for large teams where this information can be shared among all team members.

At Lyft we built a custom website that uses the JSON to show charts of code coverage trends over time, and show the source overlays in a React app that gives the generated HTML a lick of paint to make it look nicer. We use all of this for informational purposes only and not to enforce coverage minimums on anyone, but if we did all of that would have to be custom built as well.

What do other (larger) teams/companies use to visualize and distribute code coverage for their engineers in a central place?

**Comment by thedavidharris** [comment]
Our entire company uses Sonarqube, which is overall a decent tool, but it's not great with mobile development. Definitely orientied towards Java/C++. We try to use it for the minimal amount possible, and basically just use it as a historical database to track trends. We'll probably flip it soon to not even run static analysis outside of duplication checking and have it just import Swiftlint issues.

xccov has a nice JSON output that we then have a quick python script on top of to generic the format Sonarqube likes. xccov doesn't give you great branch coverage, but it's better than nothing, and llvm-cov is a bit tougher to work with. There is no way to get things like test execution reports working because of how Sonarqube wants to tie test case to specific files, and XCTest doesn't play nice with that kind of reporting.

The tooling for JUnit and Java/Kotlin is much nicer, and largely can use out of the box formats/tools there to import.

**Comment by omarzl** [comment]
We are using Slather to gather coverage and it has a decent HTML output, it can also ignore specific files or paths.
But we needed to do some modifications so it adapted better to our project.

Regarding the source overlay had you tried using:
```
xcrun llvm-cov show -instr-profile DerivedData/.../Coverage.profdata SomeApp.app/SomeApp SomeFile.swift
```
It outputs the file contents and shows which lines are missing coverage. Slather uses this output to parse it and later convert it to HTML [here](https://github.com/SlatherOrg/slather/blob/master/lib/slather/profdata_coverage_file.rb) and [here](https://github.com/SlatherOrg/slather/blob/master/lib/slather/coverage_service/html_output.rb)

**Comment by philipphofmann** [comment]
At [Sentry](https://sentry.io/) we use [Codecov](https://about.codecov.io/). It works for [Android](https://github.com/getsentry/sentry-java/blob/79634bd48656154e5a7d75a4c1423a022c079e6c/.github/workflows/build.yml#L63-L68), [iOS](https://github.com/getsentry/sentry-cocoa/blob/04beb80384b82c9d19b4b849d98fc820fcca83de/.github/workflows/test.yml#L212-L224) when using [Slather](https://github.com/getsentry/sentry-cocoa/blob/04beb80384b82c9d19b4b849d98fc820fcca83de/scripts/xcode-test.sh#L73), Flutter, and [more platforms](https://docs.codecov.com/docs/supported-languages#below-is-a-list-of-supported-languages-linking-to-their-example-repositories). I'm a bit biased, as we [bought Codecov recently](https://about.codecov.io/blog/codecov-is-joining-sentry-heres-what-you-need-to-know/), but we even used it before we bought it.
